lang/shared/src/main/scala/com/wavesplatform/lang/TypeChecker.scala

```diff
@@ -46,42 +53,64 @@ object TypeChecker {
           }
         }
 
-    case expr@Untyped.BINARY_OP(a, op, b) =>
-      (setType(ctx, EitherT.pure(a)), setType(ctx, EitherT.pure(b))).tupled
-        .subflatMap { case operands@(a, b) =>
-          val aTpe = a.tpe
-          val bTpe = b.tpe
-
-          op match {
-            case SUM_OP =>
-              if (aTpe != INT) Left(s"The first operand is expected to be INT, but got $aTpe: $a in $expr")
-              else if (bTpe != INT) Left(s"The second operand is expected to be INT, but got $bTpe: $b in $expr")
-              else Right(operands -> INT)
-
-            case GT_OP | GE_OP =>
-              if (aTpe != INT) Left(s"The first operand is expected to be INT, but got $aTpe: $a in $expr")
-              else if (bTpe != INT) Left(s"The second operand is expected to be INT, but got $bTpe: $b in $expr")
-              else Right(operands -> BOOLEAN)
-
-            case AND_OP | OR_OP =>
-              if (aTpe != BOOLEAN) Left(s"The first operand is expected to be BOOLEAN, but got $aTpe: $a in $expr")
-              else if (bTpe != BOOLEAN) Left(s"The second operand is expected to be BOOLEAN, but got $bTpe: $b in $expr")
-              else Right(operands -> BOOLEAN)
-
-            case EQ_OP =>
-              findCommonType(aTpe, bTpe)
-                .map(_ => Right(operands -> BOOLEAN))
-                .getOrElse(Left(s"Can't find common type for $aTpe and $bTpe: $a and $b in $expr"))
+    case expr @ Untyped.FUNCTION_CALL(name, args) =>
+      val value: EitherT[Coeval, String, Typed.EXPR] = ctx.functionDefs.get(name) match {
+        case Some((argTypes, resultType)) =>
+          if (args.lengthCompare(argTypes.size) != 0)
+            EitherT.fromEither[Coeval](Left(s"Function '$name' requires ${argTypes.size} arguments, but ${args.size} are provided"))
+          else {
+            import cats.instances.vector._
+            val actualArgTypes: Vector[SetTypeResult[Typed.EXPR]] = args.map(arg => setType(ctx, EitherT.pure(arg))).toVector
+            val sequencedActualArgTypes                           = actualArgTypes.sequence[SetTypeResult, Typed.EXPR].map(_.zip(argTypes))
+            sequencedActualArgTypes.subflatMap { v: Seq[(Typed.EXPR, TYPE)] =>
+              val matches = v.map {
+                case ((e, tpe)) =>
+                  findCommonType(e.tpe, tpe) match {
+                    case Some(_) => Right(e)
+                    case None    => Left(s"Types of arguments of function call '$name' don't match")
```

Comment:
Could you add to the message which types have no common parent? It is a helpful information.

Reply:
ok